src/parse_execution.cpp

```diff
@@ -770,23 +771,25 @@ end_execution_reason_t parse_execution_context_t::handle_command_not_found(
 
     // Redirect to stderr
     auto io = io_chain_t{};
-    io.append_from_specs({redirection_spec_t{STDOUT_FILENO, redirection_mode_t::fd, L"2"}}, L"");
 
-    if (function_exists(L"fish_command_not_found", *parser)) {
+    if (function_exists(L"fish_command_not_found", *parser) && parser->is_interactive()) {
         buffer = L"fish_command_not_found";
         for (const wcstring &arg : event_args) {
             buffer.push_back(L' ');
             buffer.append(escape_string(arg));
         }
-        auto prev_statuses = parser->get_last_statuses();
 
         event_t event(event_type_t::generic);
         event.desc.str_param1 = L"fish_command_not_found";
         block_t *b = parser->push_block(block_t::event_block(event));
         parser->eval(buffer, io);
         parser->pop_block(b);
-        parser->set_last_statuses(std::move(prev_statuses));
+
+        return end_execution_reason_t::error;
     } else {
+        // Redirect to stderr _if there's no handler_
+        io.append_from_specs({redirection_spec_t{STDOUT_FILENO, redirection_mode_t::fd, L"2"}},
+                             L"");
         // If we have no handler, just print it as a normal error.
         error = _(L"Unknown command:");
         if (!event_args.empty()) {
```

Comment:
This needs a newline (after the `if` which github won't allow me to mark):

```c++
error.push_back(L'\n');
```

I believe that's a pre-existing bug that this just happens to run into.